pyrsa/rdm/align.py

```diff
@@ -0,0 +1,83 @@
+"""Private implementations of align() and mean()
+"""
+import numpy as np
+from numpy import sqrt, nan, inf
+
+
+def _mean(vectors, weights=None):
+    """Weighted mean of RDM vectors, ignores nans
+
+    See :meth:`pyrsa.rdm.rdms.RDMs.mean`
+
+    Args:
+        vectors (ndarray): dissimilarity vectors of shape (nrdms, nconds)
+        weights (ndarray, optional): Same shape as vectors.
+
+    Returns:
+        ndarray: Average vector of shape (nconds,)
+    """
+    if weights is None:
+        weights = np.ones(vectors.shape)
+        weights[np.isnan(vectors)] = np.nan
+    weighted_sum = np.nansum(vectors * weights, axis=0)
+    return weighted_sum / np.nansum(weights, axis=0)
+
+
+def _ss(vectors):
+    """Sum of squares on the last dimension
+
+    Args:
+        vectors (ndarray): 1- or 2-dimensional data
+
+    Returns:
+        ndarray: the sum of squares, with an extra empty dimension
+    """
+    summed_squares = np.nansum(vectors ** 2, axis=vectors.ndim-1)
+    return np.expand_dims(summed_squares, axis=vectors.ndim-1)
+
+
+def _scale(vectors):
+    """Divide by the root sum of squares
+
+    Args:
+        vectors (ndarray): 1- or 2-dimensional data
+
+    Returns:
+        ndarray: input scaled
+    """
+    return vectors / sqrt(_ss(vectors))
+
+
+def _align(dissim, method):
+    """Align RDM vectors
+
+    See :meth:`pyrsa.rdm.rdms.RDMs.align`
+
+    Args:
+        dissim (ndarray): dissimilarity vectors, shape = (rdms, conds)
+        method (str): one of 'evidence', 'setsize' or 'simple'.
+
+    Returns:
+        (ndarray, ndarray): Tuple of the aligned dissimilarity vectors
+            and the weights used
+    """
+    n_rdms, n_conds = dissim.shape
+    if method == 'evidence':
+        weights = (dissim ** 2).clip(0.2)
```

Comment:
a hardcoded limit on the dissimilarities values seems dangerous to me

Reply:
Here's Niko's original comment from the matlab code:

> % we assume constant placement noise. therefore the dissimilarity SNR is
> % proportional to the placement distance in any particular arrangement.
> % the evidence weight of a given distance is SNR^2 (optimal filter
> % theorem). for two diametrically opposed objects at opposite sides of
> % the arena, the distance is 1 and the evidence weight is one. for a 
> % distance of .5, the SNR is .5, and the evidence weight is .25. thus it
> % takes four of those latter distances to add up to the same evidence 
> % as one of the former.
> % however, for small distances within the range of the placement noise,
> % this evidence weighting rule does not make sense: a distance of 0.1 would
> % yield an evidence weight of 0.01, requiring 100 trials to match the
> % maximum single-trial evidence weight of 1. for closely placed objects,
> % the estimate of the evidence weight itself is unreliable because the
> % placement noise dominates. so here we use the heuristic that distances
> % below 0.2 (about one fifth of the diameter) should all have the same minimum
> % evidence weight of 0.2^2=0.04, requiring about 25 trials to add up to the
> % maximum single-trial evidence weight of 1.